cylc/flow/workflow_db_mgr.py

```diff
@@ -112,27 +115,27 @@ def __init__(self, pri_d=None, pub_d=None):
             self.TABLE_ABS_OUTPUTS: []}
         self.db_updates_map = {}
 
-    def copy_pri_to_pub(self):
-        """Copy content of primary database file to public database file.
-
-        Use temporary file to ensure that we do not end up with a partial file.
-
-        """
-        temp_pub_db_file_name = None
+    def copy_pri_to_pub(self) -> None:
+        """Copy content of primary database file to public database file."""
         self.pub_dao.close()
+        # Use temporary file to ensure that we do not end up with a
+        # partial file.
+        # If an external connection is locking the old public db, it will
+        # still be connected to its inode, but should no longer affect future
+        # accesses (hopefully that process will soon recover to give up
+        # the lock).
+        temp_pub_db_fd, temp_pub_db_file_name = mkstemp(
+            prefix=self.pub_dao.DB_FILE_BASE_NAME,
+            dir=os.path.dirname(self.pub_dao.db_file_name)
+        )
+        os.close(temp_pub_db_fd)
         try:
-            self.pub_dao.conn = None  # reset connection
-            open(self.pub_dao.db_file_name, "a").close()  # touch
```

Comment:
Unhelpfully there is no explaining comment here, however, I think this open/close is more functional than it appears. This helps address potential NFS file lag which can result in a long delay between the DB being written and becoming visible across the network. The commit history might b

Reply:
Can't find any note of why this open-close/`touch` is done in the original PR #1493. It's there in the first commit that introduced the "copy pri to pub" method: https://github.com/cylc/cylc-flow/pull/1493/commits/aeb784e8520ee18e82e9abb14bc97b35348f4204#diff-6f57f386f371bc9f0ee46a01bd8ab7455793a8b84dbbbd1c98fa9a86d289621aR1380

Is the `os.rename(temp_pub_db_file_name, self.pub_dao.db_file_name)` not sufficient to create it?

Reply:
> Is the os.rename(temp_pub_db_file_name, self.pub_dao.db_file_name) not sufficient to create it?

I think the answer is an awkward combination of yes and no, yes the file exists in the metadata, no it doesn't exist in the metadata everywhere. Network filesystems are very awkward, events simultaneous in one frame aren't in another, relativistic rules apply! Matt did some investigation a while back and found that opening and closing files like this makes a difference even when `fsync` doesn't, unfortunately that's about the limit of my understanding on this.

Reply:
Ok, will revert and add some useful comments

Reply:
Thanks and sorry for these strange bits of unexplained logic, we need to get a little better at documenting the strange quirks we encounter in the fixes. I get a little nervous about introducing regressions around some of these bits of code.